File: Java/Lists/List.java
// Ryan K Lee
// Title: List.java
// Description: This file contains the class definition of a List ADT. It contains the
// necessary constructors and functions needed to create, manipulate, maintain, and delete
// a List ADT.

class List{
	private class Node{
		int data;
		Node next;
		Node prev;

		// Constructor
		Node(int data){
			this.data = data;
			next = null;
			prev = null;
		}

		public String toString() {
			return String.valueOf(data);
		}
	}

	// Fields
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int index;

	// List()
	//
	// Preconditions: none
	List(){
		front = back = null;
		cursor = null;
		length = 0;
		index = -1;
	}

    //Access Functions

    // length()
    // Returns the number of elements in this list
    // Preconditions: none
    int length(){
        return length;
    }

    // index()
    // If cursor is defined, returns the index of the cursor element, otherwise returns -1
    // Preconditions: none
    int index(){
        if (cursor != null){
            return index;
        }
        else{
            index = -1;
            return -1;
        }
    }

    // front()
    // Returns front element
    // Preconditions: length() > 0
    int front(){
        if (this.length <= 0){
            throw new RuntimeException("Error: the list is empty");
        }
        return front.data;
    }

    // back()
    // Returns back element
    // Preconditions: length() > 0
    int back(){
        if (this.length <= 0){
            throw new RuntimeException("Error: the list is empty");
        }
        return back.data;
    }

    // get()
    // Returns cursor element
    // Preconditions: length() > 0 and index() >= 0
    int get(){
        if (this.length <= 0){
            throw new RuntimeException("Error: the list is empty");
        }
        if (this.index() < 0){
            throw new RuntimeException("Error: the cursor is undefined");
        }
        return cursor.data;
    }

    // equals()
    // Returns true if and only if this List and L are the same integer sequence
    // Preconditions: None
    boolean equals(List L){
    	Node N, M;
        N = this.front;
        M = L.front;
            if (this.length() != L.length()){
                return false;
            }
            else{
                while (N != null && M != null){
                    if (N.data != M.data)
                        return false;
                    N = N.next;
                    M = M.next;
                }
                return true;
            }
    }

    //Manipulation procedures

    // clear()
    // Resets this List to its original state i.e. deletes everything in the list
    // Preconditions: none
    void clear(){
        front = back = null;
        cursor = null;
        length = 0;
        index = -1;
    }

    // moveFront()
    // If List is non-empty, places cursor under the front element, otherwise does nothing
    // Preconditions: List is not empty i.e. length() > 0
    void moveFront(){
        if (this.length > 0){
            //System.out.println("Does this work?");
            cursor = front;
            index = 0;
        }
        else{
            return;
        }
    }

    // moveBack()
    // If List is non-empty, places cursor under the back element, otherwise does nothing
    // Preconditions: List is not empty i.e. length() > 0
    void moveBack(){
        if (this.length > 0){
            cursor = back;
            index = length - 1;
        }
        else{
            return;
        }
    }

    // movePrev()
    // If cursor is defined and not at front, moves cursor one step back. If cursor is defined and at front,
    // cursor becomes undefined. If cursor is undefined, does nothing.
    // Preconditions: cursor is defined i.e. cursor != null
    void movePrev(){
        if (cursor == null){
            return;
        }
        if (cursor == front){
            cursor = null;
            index = -1;
        }
        else{
            cursor = cursor.prev;
            index--;
        }
    }

    // moveNext()
    // If cursor is defined and not at back, moves cursor one step forward. If cursor is defined and at back,
    // cursor becomes undefined. If cursor is undefined does nothing.
    // Preconditions: cursor is defined i.e. cursor != null
    void moveNext(){
        if (cursor == null){
            return;
        }
        if (cursor == back){
            cursor = null;
            index = -1;
        }
        else{
            cursor = cursor.next;
            index++;
        }
    }

    // prepend()
    // Inserts new element before the front element
    // Preconditions: none
    void prepend(int data){
        Node N = new Node(data);
        if (this.length == 0)
                front = back = N;
        else{
            front.prev = N;
            N.next = front;
            front = N;
            if (this.index >= 0){
                if (cursor == back){
                    index = -1;
                }
                else{
                    index++;
                }
            }
        }
        length++;
    }

    // append()
    // Inserts new element after the back element
    // Preconditions: none
    void append(int data){
        Node N = new Node(data);
        if (this.length == 0)
            front = back = N;
        else{
            back.next = N;
            N.prev = back;
            back = N;
        }
        length++;
    }

    // insertBefore()
    // Inserts new element before cursor
    // Preconditions: length() > 0 and index() >= 0
    void insertBefore(int data){
        if (this.length <= 0)
            throw new RuntimeException("Error: empty list");
        if (this.index() < 0)
            throw new RuntimeException("Error: the cursor is undefined");
        Node N = new Node(data);
        if (cursor == front){
            N.next = front;
            front.prev = N;
            front = N;
        }
        else{
            N.next = cursor;
            N.prev = cursor.prev;
            (cursor.prev).next = N;
            cursor.prev = N;
        }
        length++;
        index++;
    }

    // insertAfter()
    // Inserts new element after cursor
    // Preconditions: length() > 0 and index() >= 0
    void insertAfter(int data){
        if (this.length <= 0)
            throw new RuntimeException("Error: empty list");
        if (this.index() < 0)
            throw new RuntimeException("Error: the cursor is undefined");
        Node N = new Node(data);
        if (cursor == back){
            N.prev = back;
            back.next = N;
            back = N;
        }
        else{
            N.prev = cursor;
            N.next = cursor.next;
            (cursor.next).prev = N;
            cursor.next = N;
        }
        length++;
    }

    void insertOrder(int data){
    	//Node N = new Node(data);
    	Node M;
    	moveFront();
    	while (this.index() >= 0){
    		M = cursor;
    		if (data <= M.data)
    			break;
    		moveNext();
    	}
    	if (this.index() == -1){
    		append(data);
    	}
    	else{
    		insertBefore(data);
    	}
    	length++;
    }

    // deleteFront()
    // Deletes the front element
    // Preconditions: length() > 0
    void deleteFront(){
        if (this.length <= 0)
            throw new RuntimeException("Error: the list is empty");
        if (this.length > 1){
            front = front.next;
            front.prev = null;
            if (this.index == 0 || this.index == -1){
                index = -1;
            }
            else{
                index--;
            }
        }
        else{
            front = back = null;
            index = -1;
        }
        length--;
    }

    // deleteBack()
    // Deletes the back element
    // Preconditions: length() > 0
    void deleteBack(){
        if (this.length <= 0)
            throw new RuntimeException("Error: the list is empty");
        if (this.length > 1){
            back = back.prev;
            back.next = null;
            if (this.index == this.length - 1){
                index = -1;
            }
        }
        else{
            front = back = null;
            index = -1;
        }
        length--;
    }

    // delete()
    // Deletes the cursor element, making the cursor undefined
    // Preconditions: length() > 0 and index() >= 0
    void delete(){
        if (this.length <= 0)
            throw new RuntimeException("Error: the list is empty");
        if (this.index() < 0)
            throw new RuntimeException("Error: the cursor is undefined");
        Node N;
        if (cursor == front){
            front = front.next;
            front.prev = null;
            cursor = null;
            index = -1;
        }
        else if (cursor == back){
            back = back.prev;
            back.next = null;
            cursor = null;
            index = -1;
        }
        else{
            (cursor.prev).next = cursor.next;
            (cursor.next).prev = cursor.prev;
            cursor.next = null;
            cursor.prev = null;
            cursor = null;
            index = -1;
        }
        length--;
    }

    // Other methods

    // toString()
    // Overrides Object's toString method. Returns a String representation of this List consisting of a space
    // separated sequence of integers, with front on left
    // Preconditions: none
    public String toString(){
        StringBuffer s = new StringBuffer();
        Node N = front;
        while (N != null){
            s.append(N.toString());
            s.append(" ");
            N = N.next;
        }
        return new String(s);
    }

    // copy()
    // Returns a new List representing the same integer sequence as this List. The cursor in the new List
    // is undefined. This List is unchanged.
    // Preconditions: none
    List copy(){
        List L = new List();
        Node N, M;
        if (this.length > 0){
            N = this.front;
            M = new Node(N.data);
            L.front = M;
            L.length++;
            N = N.next;
            while (N != null){
                M.next = new Node(N.data);
                (M.next).prev = M;
                M = M.next;
                L.length++;
                N = N.next;
            }
        }
        return L;
    }

}

File: C/Lists/List.c
// Ryan K Lee
// Title: List.c
// Description: This file contains the structure definitions and methods needed to create,
// manipulate, maintain, and delete a List ADT
//
// Note: The cursor is a pointer the client can use to navigate the List ADT.
// The index is the element the cursor is at, ranging from 0 to length - 1.

#include<stdio.h>
#include<stdlib.h>
#include "List.h" 

// structures
typedef struct NodeObj{
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int length;
	int index;
} ListObj;

Node newNode(int data){
	Node N = malloc(sizeof(NodeObj));
	N->data = data;
	N->next = NULL;
	N->prev = NULL;

	return(N);
}

void freeNode(Node* N){
	if (N != NULL && *N != NULL){
		free(*N);
		*N = NULL;
	}
}

// Constructors and Destructors
List newList(void){
	List L = malloc(sizeof(ListObj));
	L->front = L->back = NULL;
	L->cursor = NULL;
	L->length = 0;
	L->index = -1;

	return(L);
}

void freeList(List* pL){
	if (pL != NULL && *pL != NULL){
		clear(*pL);
	}
	free(*pL);
	*pL = NULL;
}

// Access functions
int length(List L){
	if (L == NULL){
		printf("Error: length() called on NULL List reference\n");
		exit(1);
	}
	return L->length;
}

int index(List L){
	if (L == NULL){
		printf("Error: index() called on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == NULL){
		L->index = -1;
		return -1;
	}
	else{
		return L->index;
	}
}

int front(List L){
	if (L == NULL){
		printf("Error: front() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: front() called on empty list\n");
		exit(1);
	}
	return (L->front)->data;
}

int back(List L){
	if (L == NULL){
		printf("Error: back() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: back() called on empty list\n");
		exit(1);
	}
	return (L->back)->data;
}

int get(List L){
	if (L == NULL){
		printf("Error: get() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: get() called on empty list\n");
		exit(1);
	}
	if (index(L) < 0){
		printf("Error: get() called on undefined cursor\n");
		exit(1);
	}
	return (L->cursor)->data;
}

// returns 1 if equal, 0 if not equal
int equals(List A, List B){
	if (A == NULL || B == NULL){
		printf("Error: equals() called on NULL List reference\n");
		exit(1);
	}

	Node N = A->front;
	Node M = B->front;
	if (length(A) != length(B)){
		return 0;
	}
	else{
		while (N != NULL){
			if (N->data != M->data)
				return 0;
			N = N->next;
			M = M->next;
		}
		return 1;
	}

}

// Manipulation procedures
void clear(List L){
	if (L == NULL){
		printf("Error: clear() called on NULL List reference\n");
		exit(1);
	}
	// could use deleteFront or deleteBack
	while (length(L) > 0){
		Node N = L->front;
		L->front = (L->front)->next;
		N->next = NULL;
		freeNode(&N);
		L->length--;
	}
	L->cursor = NULL;
	L->length = 0;
	L->index = -1;
}

void moveFront(List L){
	if (L == NULL){
		printf("Error: moveFront() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) > 0){
        L->cursor = L->front;
        L->index = 0;
    }
    else{
        return;
    }
}

void moveBack(List L){
	if (L == NULL){
		printf("Error: moveBack() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) > 0){
        L->cursor = L->back;
        L->index = L->length - 1;
    }
    else{
        return;
    }
}

void movePrev(List L){
	if (L == NULL){
		printf("Error: movePrev() called on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == NULL){
        return;
    }
    if (L->cursor == L->front){
        L->cursor = NULL;
        L->index = -1;
    }
    else{
        L->cursor = (L->cursor)->prev;
        L->index--;
    }
}

void moveNext(List L){
	if (L == NULL){
		printf("Error: moveNext() called on NULL List reference\n");
		exit(1);
	}
	if (L->cursor == NULL){
        return;
    }
    if (L->cursor == L->back){
        L->cursor = NULL;
        L->index = -1;
    }
    else{
        L->cursor = (L->cursor)->next;
        L->index++;
    }
}

void prepend(List L, int data){
	if (L == NULL){
		printf("Error: prepend() called on NULL List reference\n");
		exit(1);
	}
	Node N = newNode(data);
        
    if (length(L) == 0){
        L->front = L->back = N;
    }
    else{
        (L->front)->prev = N;
        N->next = L->front;
        L->front = N;
        if (index(L) >= 0){
            if (index(L) == length(L) - 1){
                L->index = -1;
            }
            else{
                L->index++;
            }
        }
    }
    L->length++;
}

void append(List L, int data){
	if (L == NULL){
		printf("Error: append() called on NULL List reference\n");
		exit(1);
	}

	Node N = newNode(data);

    if (length(L) == 0){
        L->front = L->back = N;
    }
    else{
        (L->back)->next = N;
        N->prev = L->back;
        L->back = N;
    }
    L->length++;
}

void insertBefore(List L, int data){
	if (L == NULL){
		printf("Error: insertBefore() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: insertBefore() called on empty list\n");
		exit(1);
	}
	if (index(L) < 0){
		printf("Error: insertBefore() called on undefined cursor\n");
		exit(1);
	}

	Node N = newNode(data);

    if (L->cursor == L->front){
        N->next = L->front;
        (L->front)->prev = N;
        L->front = N;
    }
    else{
        N->next = L->cursor;
        N->prev = (L->cursor)->prev;
        ((L->cursor)->prev)->next = N;
        (L->cursor)->prev = N;
    }
    L->length++;
    L->index++;
}

void insertAfter(List L, int data){
	if (L == NULL){
		printf("Error: insertAfter() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: insertAfter() called on empty list\n");
		exit(1);
	}
	if (index(L) < 0){
		printf("Error: insertAfter() called on undefined cursor\n");
		exit(1);
	}

	Node N = newNode(data);

    // could find out using index and compare to length?
    if (L->cursor == L->back){
        N->prev = L->back;
        (L->back)->next = N;
        L->back = N;
    }
    else{
        N->prev = L->cursor;
        N->next = (L->cursor)->next;
        ((L->cursor)->next)->prev = N;
        (L->cursor)->next = N;
    }
    L->length++;
}

void insertOrder(List L, int data){
	if (L == NULL){
		printf("Error: insertOrder() called on NULL List reference\n");
		exit(1);
	}
	Node M;
	moveFront(L);
	while (index(L) >= 0){
		M = L->cursor;
		if (data <= M->data)
			break;
		moveNext(L);
	}
	if (index(L) == -1)
		append(L, data);
	else
		insertBefore(L, data);
	L->length++;
}

void deleteFront(List L){
	if (L == NULL){
		printf("Error: deleteFront() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: deleteFront() called on empty list\n");
		exit(1);
	}
	Node N = L->front;
    if (length(L) > 1){
        (N->next)->prev = NULL;
        L->front = (L->front)->next;
        N->next = NULL;
        if (index(L) <= 0){
            L->index = -1;
        }
        else{
            L->index--;
        }
    }
    else{
        L->front = L->back = NULL;
        L->index = -1;
    }
    L->length--;
    freeNode(&N);
}

void deleteBack(List L){
	if (L == NULL){
		printf("Error: deleteBack() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: deleteBack() called on empty list\n");
		exit(1);
	}
	Node N = L->back;
    if (length(L) > 1){
        (N->prev)->next = NULL;
        L->back = (L->back)->prev;
        N->prev = NULL;
        if (index(L) == length(L) - 1){
            L->index = -1;
        }
    }
    else{
        L->back = L->front = NULL;
        L->index = -1;
    }
    L->length--;
    freeNode(&N);
}

void delete(List L){
	if (L == NULL){
		printf("Error: delete() called on NULL List reference\n");
		exit(1);
	}
	if (length(L) <= 0){
		printf("Error: delete() called on empty list\n");
		exit(1);
	}
	if (index(L) < 0){
		printf("Error: delete() called on undefined cursor\n");
		exit(1);
	}
	Node N = NULL;

    if (L->cursor == L->front){
        N = L->front;
        L->front = (L->front)->next;
        (L->front)->prev = NULL;
        N->next = NULL;
        L->cursor = NULL;
        L->index = -1;
    }
    else if (L->cursor == L->back){
        N = L->back;
        L->back = (L->back)->prev;
        (L->back)->next = NULL;
        N->prev = NULL;
        L->cursor = NULL;
        L->index = -1;
    }
    else{
        N = L->cursor;
        (N->prev)->next = N->next;
        (N->next)->prev = N->prev;
        N->next = NULL;
        N->prev = NULL;
        L->cursor = NULL;
        L->index = -1;
    }
    freeNode(&N);
    N = NULL;
    L->length--;
}

// Miscellaneous
void printList(FILE* out, List L){
	if (out == NULL){
		printf("Error: printList() out file does not exist\n");
		exit(1);
	}
	if (L == NULL){
		printf("Error: printList() called on NULL List reference\n");
		exit(1);
	}
	Node N;
    for (N = L->front; N != NULL; N = N->next){
        fprintf(out, "%d ", N->data);
    }
    //fprintf(out, "\n");
}

List copyList(List L){
	if (L == NULL){
		printf("Error: copyList() called on NULL List reference\n");
		exit(1);
	}
	List copy = newList();
    Node N, M;

    if (length(L) > 0){
        N = L->front;
        M = newNode(N->data);
        copy->front = M;
        copy->length++;
        N = N->next;

        while (N != NULL){
            M->next = newNode(N->data);
            (M->next)->prev = M;
            M = M->next;
            copy->length++;
            N = N->next;
        }
    }
    return copy;
}